### PR TITLE
Draw timers consistently in ThreadTracks

### PR DESCRIPTION
**Pull request by karupayun**
I modified the PR and measure the new performance. Boxes are twice 
expensive than lines, at least, so send it for opinions.

Orbit has an issue when drawing timers. Sometimes empty spaces appear
between them where there shouldn't be
(http://screenshot/9D5HQVW7ZUHauTk). Is even more noticeable when at
zooming-out a new random space appear. It's reproducible since a long
time (at least in Orbit 1.65).

This is caused by the combination of 2 things:
- We are using lines and boxes to render timers.
- OpenGl doesn't guarantee pixel perfect lines.

So, In this PR we are fixing the issue for ThreadTracks. 

We are basically using the new created method to get pos and size
by extending the borders of the box and erasing the AddVerticalLine
call. This is generating a performance regression. I measured it with 
the 2-min capture presented in http://go/stadia-orbit-rendering-60fps.

With ~18k visual timers that used to be mostly lines and now they are
boxes, I found 4ms-5ms performance regression (from 131 to 135.5ms)
when sending the elements to draw from the CPU to GPU. This is
basically drawing boxes  is more expensive than lines but I think we
don't have another option.

Bug: http://b/242973688.
Test: Load a capture, play with zoom a bit.
Result: http://screenshot/6gnSk8vFuUCvyKo

### CODE REVIEW

**Comment by karupayun** [comment]
Making it draft while I'm still investigating. Feel free to give me input anyways.

**Comment by karupayun** [comment]
I made a few strict unit tests in ScopeTreeTimerData to see if we had an issue with the method, but tests are still passing, so I will set this PR ready to review as it makes sense to me.

File: src/OrbitGl/ThreadTrack.cpp [hunk 1]
```diff
@@ -436,8 +429,11 @@ void ThreadTrack::DoUpdatePrimitives(PrimitiveAssembler& primitive_assembler,
               CreatePickingUserData(primitive_assembler, *timer_info));
         }
       } else {
-        primitive_assembler.AddVerticalLine(pos, box_height, draw_data.z, color,
-                                            std::move(user_data));
+        // In this case size should be equal to 1. We will use the middle point of the start and
+        // end, so there are no empty spaces when we combine boxes with lines.
+        const float kHalfSize = 0.5f;
```
Comment (reichlfl) [inline src/OrbitGl/ThreadTrack.cpp:434]: I'd reuse the `kHalfPixel` constant above. I'm not sure about this though - you're explicitly moving the line in between two pixels, this could result in the line being drawn as 2 pixels, or randomly in of of both, or maybe even disappear (like we've seen with the horizontal lines)? I don't know if the gl specification talks about this case and it is well-defined, otherwise it might be up to the driver to decide. This sounds like a hack that hides an issue.

Are you sure this is needed?

**Review by reichlfl — COMMENTED** [review]
I think the code is a lot cleaner than before, so no concerns with that. PTAL at my comments regarding the line drawing.

File: src/OrbitGl/ThreadTrack.cpp [hunk 2]
```diff
@@ -419,13 +405,20 @@ void ThreadTrack::DoUpdatePrimitives(PrimitiveAssembler& primitive_assembler,
...
+      const float extended_end_x = ceil(end_x) + kHalfPixel;
+
+      float box_height = GetDefaultBoxHeight();
+      const Vec2 pos = {extended_start_x, world_timer_y};
+      const Vec2 size = {extended_end_x - extended_start_x, box_height};
 
-      auto timer_duration = timer_info->end() - timer_info->start();
-      if (timer_duration > draw_data.ns_per_pixel) {
+      // We are optimizing rendering by drawing the boxes with 1 pixel width as lines.
+      if (size[0] > 1.5f) {
```
[the remark this answers is not in the document]
Comment (karupayun) [reply src/OrbitGl/ThreadTrack.cpp:421]: They are float closer to integers, so I used 1.5f, but I also can use 1.f + EPS.

File: src/OrbitGl/ThreadTrack.cpp (hunk 1 above)
Comment (karupayun) [reply src/OrbitGl/ThreadTrack.cpp:434]: I completely agree that it's a hack hiding an issue because it doesn't make too much sense but without this, the issue was still there. I can add a TODO here. But I tried for several hours and this was the only combination that worked for me. As I said in the body, the other approach is just drawing rectangles and we don't need any half pixel.

File: src/OrbitGl/ThreadTrack.cpp [hunk 3]
```diff
@@ -419,13 +405,20 @@ void ThreadTrack::DoUpdatePrimitives(PrimitiveAssembler& primitive_assembler,
...
           CreatePickingUserData(primitive_assembler, *timer_info);
 
-      auto box_height = GetDefaultBoxHeight();
-      const auto [pos_x, size_x] = GetBoxPosXAndWidth(draw_data, timeline_info_, *timer_info);
-      const Vec2 pos = {pos_x, world_timer_y};
-      const Vec2 size = {size_x, box_height};
+      const float start_x = timeline_info_->GetWorldFromTick(timer_info->start());
+      const float end_x = timeline_info_->GetWorldFromTick(timer_info->end());
+      // First we are using floor and ceil to extend points to the pixel borders. After we are
+      // adding 0.5 to move the extended value to the center of the pixel.
```
[the remark this answers is not in the document]
Comment (karupayun) [reply src/OrbitGl/ThreadTrack.cpp:411]: This comment was about making users understand what we are doing, because I found that can be a bit confused, but I agree that a why is also needed.

File: src/OrbitGl/ThreadTrack.cpp (hunk 1 above)
Comment (reichlfl) [reply src/OrbitGl/ThreadTrack.cpp:434]: It works for *you*, that's what I'm concerned about. OpenGl doesn't guarantee this.

Some interesting discussion in this thread: https://stackoverflow.com/questions/45785957/opengl-line-rasterization-with-integer-endpoints - if you want to dive into the OpenGl specification: It doesn't exactly prescribe how lines need to be rendered, and they explicitly do not guarantee pixel-perfect rendering. The corner case of a line exactly in the middle of two pixels is treated by slightly overlapping the endpoints of the diamonds that the algorithm intersects with per pixel, but the way I read it this not a "must" and depends on GPU and driver implementation. They allow an error of one pixel from the reference implementation, and by specifying the line in the middle of pixels you're basically putting yourself at the mercy of the driver.

Maybe we actually want to take the 10% performance hit instead and not use lines here at all?

Comment (karupayun) [reply src/OrbitGl/ThreadTrack.cpp]: Reverted. Now we are only drawing boxes in ThreadTrack. I'm open to opinions.

**Review by florian-kuebler — APPROVED** [review]
I think this is fine. Thanks!
